File: Java/Laguna.java
public class Laguna {

    private Jugador jugador;
    private Rana[] rana;

    public Laguna(Jugador jugador){
        this.jugador = jugador;
        this.rana = new Rana[7];// por las 7 posiciones de rocas
        this.rana[0] = new Rana("v", ">");
        this.rana[1] = new Rana("v", ">");
        this.rana[2] = new Rana("v", ">");
        this.rana[3] = new Rana("-", "-");
        this.rana[4] = new Rana("r", "<");
        this.rana[5] = new Rana("r", "<");
        this.rana[6] = new Rana("r", "<");
    }

    public void comenzarJuego(){
        while(!this.hayGanador() && !this.sinMovimientos()){
            this.mostrarRocas();
            this.moverRana();
        }

        this.mostrarRocas();
        if(this.hayGanador())
            System.out.println("Felicidades!!! " + this.jugador.getNombre() + " has logrado finalizar el juego");
        else
            System.out.println("Has perdido!!! " + this.jugador.getNombre() + " sigue intentando...");

    }

    private boolean hayGanador(){
        if(!this.rana[0].vaPara("<")) return false;
        if(!this.rana[1].vaPara("<")) return false;
        if(!this.rana[2].vaPara("<")) return false;
        if(!this.rana[3].vaPara("-")) return false;
        if(!this.rana[4].vaPara(">")) return false;
        if(!this.rana[5].vaPara(">")) return false;
        if(!this.rana[6].vaPara(">")) return false;

        return true;
    }

    private boolean sinMovimientos(){
        for(int i=0; i<7; i++)
            if(!this.ranaTieneMovimiento(i)) return false;
        return true;

    }

    private boolean ranaTieneMovimiento(int i){
        if(this.rana[i].vaPara("<"))
            if(i>=2 && (this.rana[i-1].vaPara("-") || this.rana[i-2].vaPara("-"))) return true;
            else if(i==1 && this.rana[i-1].vaPara("-"))
                return true;
                else if(i==0)return true; 
        
        if(this.rana[i].vaPara(">"))
            if(i<= 4 && (this.rana[i+1].vaPara("-") || this.rana[i+2].vaPara("-"))) return true;
            else if(i==5 && this.rana[i+1].vaPara("-")) return true; 
                else if(i==6) return true;

        if(this.rana[i].vaPara("-")) return true;

        return false;
    }

    public void mostrarRocas(){
        System.out.println("\n***********************\nEstado del Juego: ");
        for(Rana rana: this.rana)
            System.out.print(rana + " ");
        System.out.println("");
        for(int i=0; i<7; i++)
            System.out.print(i + " ");

        System.out.println("\n***********************\n");
    }
    
    public void moverRana(){
        int r, m;
        Rana auxiliar;
        do{
            r = this.jugador.escogerRana();
            m = this.jugador.escogerMovimiento(this.rana[r], r);
        }while(!this.verificarMovimientoValido(r, m));

        auxiliar = this.rana[m];
        this.rana[m] = this.rana[r];
        this.rana[r] = auxiliar;
    }

    private boolean verificarMovimientoValido(int rana, int movimiento){
        return (((movimiento == rana + 1 || movimiento == rana + 2) || (movimiento == rana - 1 || movimiento == rana - 2)) && this.rana[movimiento].vaPara("-"));
    }
}

File: Python/Laguna.py
from Rana import Rana
from Jugador import Jugador

class Laguna:
    
    def __init__(self, jugador:Jugador):
        self.jugador = jugador
        self.rana = [Rana] * 7
        self.rana[0] = Rana("v", ">")
        self.rana[1] = Rana("v", ">")
        self.rana[2] = Rana("v", ">")
        self.rana[3] = Rana("-", "-")
        self.rana[4] = Rana("r", "<")
        self.rana[5] = Rana("r", "<")
        self.rana[6] = Rana("r", "<")
        
    def comenzar_juego(self):
        while not self.hay_ganador() and not self.sin_movimientos():
            self.mostrar_rocas()
            self.mover_rana()
            
        self.mostrar_rocas()
        if self.hay_ganador():
            print("Felicidades!!! " + self.jugador.get_nombre() + " has logrado finalizar el juego")
        else:
            print("Has perdido!!! " + self.jugador.get_nombre() + " sigue intentando...")
    
    def hay_ganador(self):
        if not self.rana[0].va_para("<"): return False
        if not self.rana[1].va_para("<"): return False
        if not self.rana[2].va_para("<"): return False
        if not self.rana[3].va_para("-"): return False
        if not self.rana[4].va_para(">"): return False
        if not self.rana[5].va_para(">"): return False
        if not self.rana[6].va_para(">"): return False
        
        return True
    
    def sin_movimientos(self):
        for i in range(7):
            if not self.rana_tiene_movimiento(i): return False
        return True
    
    def rana_tiene_movimiento(self, i:int):
        
        if self.rana[i].va_para("<") and ((i>=2 and (self.rana[i-1].va_para("-") or self.rana[i-2].va_para("-"))) or (i == 1 and self.rana[i-1].va_para("-")) or i== 0): 
            return True
        
        if self.rana[i].va_para(">") and ((i>=4 and (self.rana[i+1].va_para("-") or self.rana[i+2].va_para("-"))) or (i == 5 and self.rana[i+1].va_para("-")) or i==6): 
            return True
        if self.rana[i].va_para("-"):
            return True
        
        return False
    
    def mostrar_rocas(self):
        print("\n***********************\nEstado del Juego: ")
        for rana in self.rana:
            print(rana, end=" ")
        print("")
        for i in range(7):
            print(i, end=" ")
        print("\n***********************\n")
        
    def mover_rana(self):
        r = self.jugador.escoger_rana()
        m = self.jugador.escoger_movimiento(self.rana[r], r)
        while not self.verificar_movimiento_valido(r,m):
            r = self.jugador.escoger_rana()
            m = self.jugador.escoger_movimiento(self.rana[r], r)
            
        auxiliar = self.rana[m]
        self.rana[m] = self.rana[r]
        self.rana[r] = auxiliar
        
    def verificar_movimiento_valido(self, rana:int, movimiento:int):
        return ((movimiento == rana + 1 or movimiento == rana +2) or (movimiento == rana - 1 or movimiento == rana - 2)) and self.rana[movimiento].va_para("-")
            